File: backend/app/processing/folder_sync.py
import os
import shutil
import uuid
from pathlib import Path
from typing import Dict, Any, List
from app.core.config import settings
from app.database.session import db_get, db_run
from app.processing.queue import enqueue_source_processing

SUPPORTED_EXTENSIONS = {
    'pdf', 'docx', 'doc', 'txt', 'md', 'csv', 'xlsx', 'xls', 'json', 'xml', 'html',
    'jpg', 'jpeg', 'png', 'webp', 'gif', 'bmp',
    'mp4', 'avi', 'mov', 'mkv', 'webm',
    'mp3', 'wav', 'm4a', 'ogg', 'flac', 'aac'
}

async def sync_knowledge_folder() -> Dict[str, Any]:
    from app.database.seed import seed_system_user
    seed_system_user()

    kb_dir = settings.resolved_kb_dir
    upload_dir = settings.resolved_upload_dir
    kb_dir.mkdir(parents=True, exist_ok=True)
    upload_dir.mkdir(parents=True, exist_ok=True)

    synced_files = []
    skipped_files = []
    errors = []

    files = list(kb_dir.glob("*"))
    for file_path in files:
        if file_path.is_dir() or file_path.name.startswith("."):
            continue

        ext = file_path.suffix.lstrip(".").lower()
        if ext not in SUPPORTED_EXTENSIONS:
            skipped_files.append({"filename": file_path.name, "reason": "unsupported_extension"})
            continue

        original_name = file_path.name
        existing = db_get(
            "SELECT id, processing_status FROM knowledge_sources WHERE original_name = ?",
            (original_name,)
        )

        if existing and existing["processing_status"] in ("completed", "processing", "indexing"):
            skipped_files.append({"filename": original_name, "reason": "already_indexed", "id": existing["id"]})
            continue

        source_id = existing["id"] if existing else str(uuid.uuid4())
        dest_filename = f"{source_id}.{ext}"
        dest_path = upload_dir / dest_filename

        try:
            shutil.copy2(file_path, dest_path)
            file_size = file_path.stat().st_size

            if existing:
                db_run(
                    "UPDATE knowledge_sources SET file_path = ?, file_size = ?, processing_status = 'uploaded' WHERE id = ?",
                    (str(dest_path), file_size, source_id)
                )
            else:
                db_run(
                    """INSERT INTO knowledge_sources (
                        id, filename, original_name, file_type, file_path, file_size, uploaded_by, processing_status
                    ) VALUES (?, ?, ?, ?, ?, ?, 'system', 'uploaded')""",
                    (source_id, dest_filename, original_name, ext, str(dest_path), file_size)
                )

            await enqueue_source_processing(source_id)
            synced_files.append({"id": source_id, "originalName": original_name, "fileType": ext, "status": "queued"})
        except Exception as e:
            errors.append({"filename": original_name, "error": str(e)})

    return {
        "scanned": len(files),
        "synced": len(synced_files),
        "skipped": len(skipped_files),
        "errors": len(errors),
        "syncedFiles": synced_files,
        "skippedFiles": skipped_files,
        "errorDetails": errors,
    }

File: workspace/src/lib/knowledge/folder-sync.ts
// src/lib/knowledge/folder-sync.ts
// Automatic Knowledge Base Folder Ingestion & Sync Engine
// Automatically scans the `knowledge_base` folder on server start & on demand,
// ensuring all reference materials (PDF, DOCX, Video, Audio, Image) are permanently preserved and indexed in SQLite.

import fs from 'fs';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { dbGet, dbAll, dbRun, saveDb, dbTransaction } from '@/lib/db/client';
import { processDocument } from './document-processor';
import { processImage } from './image-processor';
import { processVideo } from './video-processor';
import { processAudio } from './audio-processor';

export const KB_DIR = path.join(process.cwd(), 'knowledge_base');
export const UPLOAD_DIR = path.join(process.cwd(), 'uploads');

const SUPPORTED_EXTENSIONS = new Set([
  // Documents
  'pdf', 'docx', 'doc', 'txt', 'md', 'markdown', 'csv', 'xlsx', 'xls', 'json', 'xml', 'html', 'pptx',
  // Images
  'jpg', 'jpeg', 'png', 'webp', 'gif', 'bmp',
  // Videos
  'mp4', 'avi', 'mov', 'mkv', 'webm',
  // Audio
  'mp3', 'wav', 'm4a', 'ogg', 'flac', 'aac',
]);

let isSyncing = false;

/**
 * Ensures the `knowledge_base` and `uploads` directories exist.
 */
export function ensureKnowledgeFoldersExist() {
  if (!fs.existsSync(KB_DIR)) {
    fs.mkdirSync(KB_DIR, { recursive: true });
  }
  if (!fs.existsSync(UPLOAD_DIR)) {
    fs.mkdirSync(UPLOAD_DIR, { recursive: true });
  }
}

/**
 * Automatically scans the `knowledge_base/` folder and indexes any new or changed files into SQLite.
 */
export async function syncKnowledgeFolder(): Promise<{
  totalFiles: number;
  newlyIndexed: number;
  alreadyIndexed: number;
  errors: number;
  syncedFiles: string[];
}> {
  ensureKnowledgeFoldersExist();

  if (isSyncing) {
    console.log('[FolderSync] Sync already in progress, skipping concurrent run.');
    return { totalFiles: 0, newlyIndexed: 0, alreadyIndexed: 0, errors: 0, syncedFiles: [] };
  }

  isSyncing = true;
  console.log('[FolderSync] Starting knowledge_base folder scan at:', KB_DIR);

  let newlyIndexed = 0;
  let alreadyIndexed = 0;
  let errors = 0;
  const syncedFiles: string[] = [];

  try {
    const files = fs.readdirSync(KB_DIR);
    const validFiles = files.filter((f) => {
      const ext = path.extname(f).slice(1).toLowerCase();
      const stat = fs.statSync(path.join(KB_DIR, f));
      return stat.isFile() && SUPPORTED_EXTENSIONS.has(ext);
    });

    // Get system/admin user ID
    const adminUser = dbGet<{ id: string }>('SELECT id FROM users LIMIT 1');
    const uploaderId = adminUser?.id || 'admin-system';

    for (const fileName of validFiles) {
      const filePath = path.join(KB_DIR, fileName);
      const stat = fs.statSync(filePath);
      const ext = path.extname(fileName).slice(1).toLowerCase();

      // Check if already indexed in database with completed status
      const existing = dbGet<{
        id: string;
        processing_status: string;
        chunk_count: number;
        transcript_count: number;
        face_count: number;
        file_size: number;
      }>('SELECT id, processing_status, chunk_count, transcript_count, face_count, file_size FROM knowledge_sources WHERE original_name = ?', [fileName]);

      const isFullyProcessed =
        existing &&
        existing.processing_status === 'completed' &&
        (existing.chunk_count > 0 || existing.transcript_count > 0 || existing.face_count > 0 || ['jpg', 'png', 'jpeg', 'webp'].includes(ext));

      if (isFullyProcessed) {
        alreadyIndexed++;
        syncedFiles.push(fileName);
        continue;
      }

      console.log(`[FolderSync] Indexing new knowledge asset: "${fileName}" (${(stat.size / 1024).toFixed(1)} KB)...`);

      const sourceId = existing ? existing.id : uuidv4();
      const safeFilename = `${sourceId}.${ext}`;
      const destUploadPath = path.join(UPLOAD_DIR, safeFilename);

      // Copy to uploads folder for web streaming/viewing
      try {
        if (!fs.existsSync(destUploadPath) || fs.statSync(destUploadPath).size !== stat.size) {
          fs.copyFileSync(filePath, destUploadPath);
        }
      } catch (err) {
        console.warn(`[FolderSync] Notice copying to uploads:`, err);
      }

      if (!existing) {
        dbRun(`
          INSERT INTO knowledge_sources (
            id, filename, original_name, file_type, file_path, file_size, uploaded_by, processing_status
          ) VALUES (?, ?, ?, ?, ?, ?, ?, 'indexing')
        `, [sourceId, safeFilename, fileName, ext, destUploadPath, stat.size, uploaderId]);
      } else {
        dbRun('UPDATE knowledge_sources SET processing_status = ? WHERE id = ?', ['indexing', sourceId]);
      }

      saveDb(true);

      // Process and index based on media type
      try {
        const isDocument = ['pdf', 'docx', 'doc', 'txt', 'md', 'csv', 'xlsx', 'xls', 'json', 'xml', 'html', 'markdown', 'pptx'].includes(ext);
        const isImage = ['jpg', 'jpeg', 'png', 'webp', 'gif', 'bmp'].includes(ext);
        const isVideo = ['mp4', 'avi', 'mov', 'mkv', 'webm'].includes(ext);
        const isAudio = ['mp3', 'wav', 'm4a', 'ogg', 'flac', 'aac'].includes(ext);

        if (isDocument) {
          await processDocument(sourceId, destUploadPath, ext, fileName);
        } else if (isImage) {
          await processImage(sourceId, destUploadPath, ext, fileName);
        } else if (isVideo) {
          await processVideo(sourceId, destUploadPath, ext, fileName);
        } else if (isAudio) {
          await processAudio(sourceId, destUploadPath, ext, fileName);
        }

        dbRun("UPDATE knowledge_sources SET processing_status = 'completed' WHERE id = ?", [sourceId]);
        saveDb(true);
        newlyIndexed++;
        syncedFiles.push(fileName);
        console.log(`[FolderSync] Successfully indexed: "${fileName}"`);
      } catch (err) {
        errors++;
        const msg = err instanceof Error ? err.message : String(err);
        console.error(`[FolderSync] Error indexing "${fileName}":`, err);
        dbRun("UPDATE knowledge_sources SET processing_status = 'failed', error_message = ? WHERE id = ?", [msg, sourceId]);
        saveDb(true);
      }
    }

    console.log(`[FolderSync] Sync complete. Total: ${validFiles.length} files (New: ${newlyIndexed}, Already Indexed: ${alreadyIndexed}, Errors: ${errors})`);
    return {
      totalFiles: validFiles.length,
      newlyIndexed,
      alreadyIndexed,
      errors,
      syncedFiles,
    };
  } finally {
    isSyncing = false;
  }
}
